File: src/cn/yiiguxing/plugin/translate/Styles.java
package cn.yiiguxing.plugin.translate;

import cn.yiiguxing.plugin.translate.model.BasicExplain;
import cn.yiiguxing.plugin.translate.model.QueryResult;
import cn.yiiguxing.plugin.translate.model.WebExplain;
import cn.yiiguxing.plugin.translate.ui.PhoneticButton;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.ui.JBColor;
import com.intellij.util.Consumer;
import com.intellij.util.ui.JBUI;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.text.*;
import java.awt.event.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 文本样式
 */
@SuppressWarnings("SpellCheckingInspection")
public final class Styles {

    private static final Logger LOGGER = Logger.getInstance("#" + Styles.class.getCanonicalName());

    private static final SimpleAttributeSet ATTR_QUERY = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_EXPLAIN = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_PRE_EXPLAINS = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_EXPLAINS = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_EXPLAINS_HOVER = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_WEB_EXPLAIN_TITLE = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_WEB_EXPLAIN_KEY = new SimpleAttributeSet();
    private static final SimpleAttributeSet ATTR_WEB_EXPLAIN_VALUES = new SimpleAttributeSet();

    private static final int QUERY_FONT_SIZE = 19;
    private static final int PRE_EXPLAINS_FONT_SIZE = 16;

    private static final int EXPLAINS_FONT_SIZE = 16;

    private static final Pattern PATTERN_WORD = Pattern.compile("[a-zA-Z]+");


    static {
        StyleConstants.setItalic(ATTR_QUERY, true);
        StyleConstants.setBold(ATTR_QUERY, true);
        StyleConstants.setForeground(ATTR_QUERY, new JBColor(0xFFEE6000, 0xFFCC7832));

        StyleConstants.setForeground(ATTR_EXPLAIN, new JBColor(0xFF3E7EFF, 0xFF8CBCE1));

        StyleConstants.setItalic(ATTR_PRE_EXPLAINS, true);
        StyleConstants.setForeground(ATTR_PRE_EXPLAINS, new JBColor(0xFF7F0055, 0xFFEAB1FF));
        StyleConstants.setFontSize(ATTR_PRE_EXPLAINS, JBUI.scaleFontSize(16));

        StyleConstants.setForeground(ATTR_EXPLAINS, new JBColor(0xFF170591, 0xFFFFC66D));
        StyleConstants.setFontSize(ATTR_EXPLAINS, JBUI.scaleFontSize(16));

        StyleConstants.setForeground(ATTR_EXPLAINS_HOVER, new JBColor(0xA60EFF, 0xDF531F));
        StyleConstants.setFontSize(ATTR_EXPLAINS_HOVER, JBUI.scaleFontSize(16));

        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_TITLE, new JBColor(0xFF707070, 0xFF808080));
        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_KEY, new JBColor(0xFF4C4C4C, 0xFF77B767));
        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_VALUES, new JBColor(0xFF707070, 0xFF6A8759));
    }

    private Styles() {
    }

    private static void setMouseListeners(@NotNull JTextPane textPane) {
        for (MouseListener listener : textPane.getMouseListeners()) {
            if (listener instanceof ClickableStyleListener) {
                textPane.removeMouseListener(listener);
            }
        }
        for (MouseMotionListener listener : textPane.getMouseMotionListeners()) {
            if (listener instanceof ClickableStyleListener) {
                textPane.removeMouseMotionListener(listener);
            }
        }

        MouseAdapter listener = new ClickableStyleListener();
        textPane.addMouseListener(listener);
        textPane.addMouseMotionListener(listener);
    }

    public static void insertStylishResultText(@NotNull final JTextPane textPane,
                                               @NotNull QueryResult result,
                                               @Nullable OnTextClickListener explainsClickListener) {
        setMouseListeners(textPane);

        final StyledDocument document = (StyledDocument) textPane.getDocument();

        try {
            document.remove(0, document.getLength());
        } catch (BadLocationException e) {
            e.printStackTrace();
            return;
        }

        insertHeader(document, result);

        BasicExplain basicExplain = result.getBasicExplain();
        if (basicExplain != null) {
            insertExplain(textPane, document, basicExplain.getExplains(), true, explainsClickListener);
        } else {
            insertExplain(textPane, document, result.getTranslation(), false, explainsClickListener);
        }

        WebExplain[] webExplains = result.getWebExplains();
        insertWebExplain(document, webExplains);

        if (document.getLength() < 1)
            return;

        try {
            int offset = document.getLength() - 1;
            String text = document.getText(offset, 1);
            if (text.charAt(0) == '\n') {
                document.remove(offset, 1);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    // 不能静态设置，否则scale改变时不能即时更新
    @NotNull
    private static MutableAttributeSet updateFontSize(@NotNull MutableAttributeSet attr, int size) {
        StyleConstants.setFontSize(attr, JBUI.scaleFontSize(size));
        return attr;
    }

    private static void insertHeader(@NotNull Document document, QueryResult result) {
        String query = result.getQuery();

        try {
            if (!Utils.isEmptyOrBlankString(query)) {
                query = query.trim();
                document.insertString(document.getLength(),
                        Character.toUpperCase(query.charAt(0)) + query.substring(1) + "\n",
                        updateFontSize(ATTR_QUERY, QUERY_FONT_SIZE));
            }

            BasicExplain be = result.getBasicExplain();
            if (be != null) {
                boolean hasPhonetic = false;

                String phoUK = be.getPhoneticUK();
                if (!Utils.isEmptyOrBlankString(phoUK)) {
                    insertPhonetic(document, result.getQuery(), phoUK, Speech.Phonetic.UK);
                    hasPhonetic = true;
                }

                String phoUS = be.getPhoneticUS();
                if (!Utils.isEmptyOrBlankString(phoUS)) {
                    insertPhonetic(document, result.getQuery(), phoUS, Speech.Phonetic.US);
                    hasPhonetic = true;
                }

                String pho = be.getPhonetic();
                if (!Utils.isEmptyOrBlankString(pho) && !hasPhonetic) {
                    document.insertString(document.getLength(), "[" + pho + "]", ATTR_EXPLAIN);
                    hasPhonetic = true;
                }

                if (hasPhonetic) {
                    document.insertString(document.getLength(), "\n", null);
                }
            }

            document.insertString(document.getLength(), "\n", null);
        } catch (BadLocationException e) {
            LOGGER.error("insertHeader ", e);
        }
    }

    private static void insertPhonetic(@NotNull Document document,
                                       @NotNull final String query,
                                       @NotNull String phoneticText,
                                       @NotNull final Speech.Phonetic phonetic) throws BadLocationException {
        String insert;
        if (phonetic == Speech.Phonetic.UK) {
            insert = "英[";
        } else {
            insert = "美[";
        }
        insert += phoneticText + "]";
        document.insertString(document.getLength(), insert, ATTR_EXPLAIN);

        SimpleAttributeSet attr = new SimpleAttributeSet();
        StyleConstants.setComponent(attr, new PhoneticButton(new Consumer<MouseEvent>() {
            @Override
            public void consume(MouseEvent mouseEvent) {
                if (mouseEvent.getClickCount() == 1) {
                    Speech.toSpeech(query, phonetic);
                }
            }
        }));
        document.insertString(document.getLength(), " ", attr);
    }

    private static void insertExplain(@NotNull final JTextPane textPane,
                                      @NotNull StyledDocument doc,
                                      @Nullable String[] explains,
                                      boolean splitLabel,
                                      @Nullable OnTextClickListener explainsClickListener) {
        if (explains == null || explains.length == 0)
            return;

        MutableAttributeSet attrPre = updateFontSize(ATTR_PRE_EXPLAINS, PRE_EXPLAINS_FONT_SIZE);
        MutableAttributeSet attr = updateFontSize(ATTR_EXPLAINS, EXPLAINS_FONT_SIZE);
        try {
            for (String exp : explains) {
                if (Utils.isEmptyOrBlankString(exp))
                    continue;

                if (splitLabel) {
                    String[] splits = Utils.splitExplain(exp);
                    if (splits[0] != null) {
                        doc.insertString(doc.getLength(), splits[0] + " ", attrPre);
                        exp = splits[1];
                    }

                    final int offset = doc.getLength();
                    doc.insertString(offset, exp + "\n", attr);

                    Matcher wordMatcher = PATTERN_WORD.matcher(exp);
                    String text;
                    int start;
                    ClickableStyle style;
                    while (wordMatcher.find()) {
                        text = wordMatcher.group();
                        start = wordMatcher.start() + offset;
                        style = new ClickableStyle(textPane, text, start, explainsClickListener);
                        doc.setCharacterAttributes(start, text.length(), setClickableStyle(attr, style), true);
                    }
                } else {
                    doc.insertString(doc.getLength(), exp + "\n", attr);
                }
            }

            doc.insertString(doc.getLength(), "\n", null);
        } catch (BadLocationException e) {
            LOGGER.error("insertExplain ", e);
        }
    }

    private static void insertWebExplain(Document doc, WebExplain[] webExplains) {
        if (webExplains == null || webExplains.length == 0)
            return;

        try {
            doc.insertString(doc.getLength(), "网络释义:\n", ATTR_WEB_EXPLAIN_TITLE);

            for (WebExplain webExplain : webExplains) {
                doc.insertString(doc.getLength(), webExplain.getKey(), ATTR_WEB_EXPLAIN_KEY);
                doc.insertString(doc.getLength(), " -", null);

                String[] values = webExplain.getValues();
                for (int i = 0; i < values.length; i++) {
                    doc.insertString(doc.getLength(), " " + values[i] + (i < values.length - 1 ? ";" : ""),
                            ATTR_WEB_EXPLAIN_VALUES);
                }
                doc.insertString(doc.getLength(), "\n", null);
            }

        } catch (BadLocationException e) {
            LOGGER.error("insertWebExplain ", e);
        }
    }

    private static MutableAttributeSet setClickableStyle(MutableAttributeSet attrSet, ClickableStyle style) {
        attrSet = (MutableAttributeSet) attrSet.copyAttributes();
        attrSet.addAttribute(ClickableStyle.class, style);
        return attrSet;
    }

    public interface OnTextClickListener {
        void onTextClick(@NotNull JTextPane textPane, @NotNull String text);
    }

    private static final class ClickableStyle {
        private final JTextPane mTextPane;
        private final String mText;
        private final int mStartOffset;
        private final OnTextClickListener mListener;

        private boolean mHover;

        public ClickableStyle(@NotNull JTextPane textPane, @NotNull String text, int startOffset,
                              @Nullable OnTextClickListener listener) {
            this.mTextPane = textPane;
            this.mText = text;
            this.mStartOffset = startOffset;
            this.mListener = listener;
        }

        void performClick() {
            if (mListener != null) {
                mListener.onTextClick(mTextPane, mText);
            }
        }

        void onHover() {
            if (!mHover) {
                StyledDocument document = (StyledDocument) mTextPane.getDocument();
                MutableAttributeSet attr = updateFontSize(ATTR_EXPLAINS_HOVER, EXPLAINS_FONT_SIZE);
                attr = setClickableStyle(attr, this);
                document.setCharacterAttributes(mStartOffset, mText.length(), attr, true);

                mHover = true;
            }
        }

        void clearHover() {
            if (mHover) {
                StyledDocument document = (StyledDocument) mTextPane.getDocument();
                MutableAttributeSet attr = updateFontSize(ATTR_EXPLAINS, EXPLAINS_FONT_SIZE);
                attr = setClickableStyle(attr, this);
                document.setCharacterAttributes(mStartOffset, mText.length(), attr, true);

                mHover = false;
            }
        }

    }

    private static final class ClickableStyleListener extends MouseAdapter {
        private ClickableStyle mLastHover;

        @Override
        public void mouseClicked(MouseEvent e) {
            if ((e.getModifiers() & InputEvent.BUTTON1_MASK) == 0 || e.getClickCount() > 1)
                return;

            ClickableStyle clickableStyle = getClickableStyle(e);
            if (clickableStyle != null) {
                clickableStyle.performClick();
            }
        }

        private ClickableStyle getClickableStyle(MouseEvent e) {
            JTextPane textPane = (JTextPane) e.getComponent();
            StyledDocument document = (StyledDocument) textPane.getDocument();

            Element ele = document.getCharacterElement(textPane.viewToModel(e.getPoint()));
            MutableAttributeSet as = (MutableAttributeSet) ele.getAttributes();
            return (ClickableStyle) as.getAttribute(ClickableStyle.class);
        }

        @Override
        public void mouseMoved(MouseEvent e) {
            final ClickableStyle lastHover = mLastHover;
            final ClickableStyle hover = getClickableStyle(e);

            if (lastHover != hover) {
                mLastHover = hover;

                if (lastHover != null) {
                    lastHover.clearHover();
                }
                if (hover != null) {
                    hover.onHover();
                }
            }
        }

        @Override
        public void mouseExited(MouseEvent e) {
            if (mLastHover != null) {
                mLastHover.clearHover();
                mLastHover = null;
            }
        }
    }

}

File: src/main/kotlin/cn/yiiguxing/plugin/translate/Styles.kt
package cn.yiiguxing.plugin.translate

import cn.yiiguxing.plugin.translate.model.QueryResult
import cn.yiiguxing.plugin.translate.model.WebExplain
import cn.yiiguxing.plugin.translate.tts.TextToSpeech
import cn.yiiguxing.plugin.translate.ui.PhoneticButton
import cn.yiiguxing.plugin.translate.util.TranslationResultUtils
import cn.yiiguxing.plugin.translate.util.appendString
import cn.yiiguxing.plugin.translate.util.clear
import cn.yiiguxing.plugin.translate.util.trimEnd
import com.intellij.openapi.diagnostic.Logger
import com.intellij.ui.JBColor
import com.intellij.util.Consumer
import java.awt.Font
import java.awt.event.InputEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.regex.Pattern
import javax.swing.JTextPane
import javax.swing.text.*

/**
 * 文本样式
 */
object Styles {

    private val LOGGER = Logger.getInstance("#" + Styles::class.java.canonicalName)

    private val ATTR_QUERY = SimpleAttributeSet()
    private val ATTR_EXPLAIN_BASE = SimpleAttributeSet()
    private val ATTR_EXPLAIN = SimpleAttributeSet()
    private val ATTR_PRE_EXPLAINS = SimpleAttributeSet()
    private val ATTR_EXPLAINS = SimpleAttributeSet()
    private val ATTR_EXPLAINS_HOVER = SimpleAttributeSet()
    private val ATTR_WEB_EXPLAIN_TITLE = SimpleAttributeSet()
    private val ATTR_WEB_EXPLAIN_KEY = SimpleAttributeSet()
    private val ATTR_WEB_EXPLAIN_VALUES = SimpleAttributeSet()

    private const val QUERY_FONT_SCALE = 1.35f
    private const val PRE_EXPLAINS_FONT_SCALE = 1.15f
    private const val EXPLAINS_FONT_SCALE = 1.15f

    private val PATTERN_WORD = Pattern.compile("[a-zA-Z]+")

    init {
        ATTR_QUERY.let {
            StyleConstants.setItalic(it, true)
            StyleConstants.setBold(it, true)
            StyleConstants.setForeground(it, JBColor(0xFFEE6000.toInt(), 0xFFCC7832.toInt()))
        }

        val fg = JBColor(0xFF3E7EFF.toInt(), 0xFF8CBCE1.toInt())
        StyleConstants.setForeground(ATTR_EXPLAIN_BASE, fg)
        StyleConstants.setForeground(ATTR_EXPLAIN, fg)

        StyleConstants.setItalic(ATTR_PRE_EXPLAINS, true)
        StyleConstants.setForeground(ATTR_PRE_EXPLAINS, JBColor(0xFF7F0055.toInt(), 0xFFEAB1FF.toInt()))

        StyleConstants.setForeground(ATTR_EXPLAINS, JBColor(0xFF170591.toInt(), 0xFFFFC66D.toInt()))

        StyleConstants.setForeground(ATTR_EXPLAINS_HOVER, JBColor(0xA60EFF, 0xDF531F))

        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_TITLE, JBColor(0xFF707070.toInt(), 0xFF808080.toInt()))
        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_KEY, JBColor(0xFF4C4C4C.toInt(), 0xFF77B767.toInt()))
        StyleConstants.setForeground(ATTR_WEB_EXPLAIN_VALUES, JBColor(0xFF707070.toInt(), 0xFF6A8759.toInt()))
    }

    private fun setMouseListeners(textPane: JTextPane) {
        textPane.mouseListeners
                .filterIsInstance<ClickableStyleListener>()
                .forEach { textPane.removeMouseListener(it) }
        textPane.mouseMotionListeners
                .filterIsInstance<ClickableStyleListener>()
                .forEach { textPane.removeMouseMotionListener(it) }

        val listener = ClickableStyleListener()
        textPane.addMouseListener(listener)
        textPane.addMouseMotionListener(listener)
    }

    fun insertStylishResultText(textPane: JTextPane,
                                result: QueryResult,
                                explainsClickListener: OnTextClickListener?) {
        setMouseListeners(textPane)

        val document = textPane.styledDocument
        document.clear()

        val query = result.query
        if (query == null || query.isBlank()) {
            document.appendString("Noting to show.")
            return
        }

        insertHeader(textPane, query, result)

        val basicExplain = result.basicExplain
        if (basicExplain != null) {
            insertExplain(textPane, document, basicExplain.explains, true, explainsClickListener)
        } else {
            insertExplain(textPane, document, result.translation, false, explainsClickListener)
        }

        val webExplains = result.webExplains
        insertWebExplain(document, webExplains)

        document.trimEnd()
    }

    // 不能静态设置，否则scale改变时不能即时更新
    private fun updateFontSize(attr: MutableAttributeSet, font: Font, scale: Float): MutableAttributeSet = attr.apply {
        StyleConstants.setFontSize(this, Math.round(font.size * scale))
    }

    enum class Phonetic(val value: Int) {
        /**
         * 英式发音
         */
        UK(1),
        /**
         * 美式发音
         */
        US(2)
    }

    private fun insertHeader(textPane: JTextPane, query: String, result: QueryResult) {
        val document = textPane.document

        try {
            val title = query.trim { it <= ' ' }
            document.appendString(title.capitalize() + "\n",
                    updateFontSize(ATTR_QUERY, textPane.font, QUERY_FONT_SCALE))


            result.basicExplain?.let { (pho, phoUK, phoUS) ->
                var hasPhonetic = false

                if (!phoUK.isNullOrBlank()) {
                    insertPhonetic(document, title, phoUK!!, Phonetic.UK)
                    hasPhonetic = true
                }
                if (!phoUS.isNullOrBlank()) {
                    insertPhonetic(document, title, phoUS!!, Phonetic.US)
                    hasPhonetic = true
                }

                if (!hasPhonetic && !pho.isNullOrBlank()) {
                    document.appendString("[$pho]", ATTR_EXPLAIN)
                    hasPhonetic = true
                }

                if (hasPhonetic) {
                    document.appendString("\n")
                }
            }

            document.appendString("\n")
        } catch (e: BadLocationException) {
            LOGGER.error("insertHeader ", e)
        }
    }

    private fun insertPhonetic(document: Document,
                               query: String,
                               phoneticText: String,
                               phonetic: Phonetic) {
        document.appendString(if (phonetic === Phonetic.UK) "英[" else "美[", ATTR_EXPLAIN_BASE)

        val settings = Settings.instance
        val fontFamily = settings.phoneticFontFamily
        if (!settings.isOverrideFont || fontFamily.isNullOrBlank()) {
            ATTR_EXPLAIN.removeAttribute(StyleConstants.FontFamily)
        } else {
            StyleConstants.setFontFamily(ATTR_EXPLAIN, fontFamily)
        }

        document.appendString(phoneticText, ATTR_EXPLAIN)
        document.appendString("]", ATTR_EXPLAIN_BASE)

        val attr = SimpleAttributeSet()
        StyleConstants.setComponent(attr, PhoneticButton(Consumer { mouseEvent ->
            if (mouseEvent.clickCount == 1) {
                TextToSpeech.INSTANCE.speak(query)
            }
        }))
        document.appendString(" ", attr)
    }

    private fun insertExplain(textPane: JTextPane,
                              doc: StyledDocument,
                              explains: Array<String>?,
                              splitLabel: Boolean,
                              explainsClickListener: OnTextClickListener?) {
        if (explains == null || explains.isEmpty())
            return

        val attrPre = updateFontSize(ATTR_PRE_EXPLAINS, textPane.font, PRE_EXPLAINS_FONT_SCALE)
        val attr = updateFontSize(ATTR_EXPLAINS, textPane.font, EXPLAINS_FONT_SCALE)
        try {
            explains.filter { it.isNotBlank() }.forEach {
                if (splitLabel) {
                    val (wordClass, explain) = TranslationResultUtils.splitExplain(it)

                    wordClass?.let {
                        doc.appendString("$it ", attrPre)
                    }

                    val offset = doc.length
                    doc.appendString("$explain\n", attr)

                    val wordMatcher = PATTERN_WORD.matcher(explain)
                    while (wordMatcher.find()) {
                        val text = wordMatcher.group()
                        val start = wordMatcher.start() + offset
                        val style = ClickableStyle(textPane, text, start, explainsClickListener)
                        doc.setCharacterAttributes(start, text.length, setClickableStyle(attr, style), true)
                    }
                } else {
                    doc.appendString("$it\n", attr)
                }
            }
            doc.appendString("\n")
        } catch (e: BadLocationException) {
            LOGGER.error("insertExplain ", e)
        }
    }

    private fun insertWebExplain(doc: Document, webExplains: Array<WebExplain>?) {
        if (webExplains == null || webExplains.isEmpty())
            return

        try {
            doc.appendString("网络释义:\n", ATTR_WEB_EXPLAIN_TITLE)

            for ((key, values) in webExplains) {
                if (key == null) continue

                doc.appendString(key, ATTR_WEB_EXPLAIN_KEY)

                if (values != null && values.isNotEmpty()) {
                    doc.appendString(" -")

                    for (i in values.indices) {
                        doc.appendString(
                                " " + values[i] + if (i < values.size - 1) ";" else "",
                                ATTR_WEB_EXPLAIN_VALUES
                        )
                    }
                }

                doc.appendString("\n")
            }
        } catch (e: BadLocationException) {
            LOGGER.error("insertWebExplain ", e)
        }
    }

    private fun setClickableStyle(attrSet: MutableAttributeSet, style: ClickableStyle): MutableAttributeSet {
        return (attrSet.copyAttributes() as MutableAttributeSet).apply {
            addAttribute(ClickableStyle::class.java, style)
        }
    }

    interface OnTextClickListener {
        fun onTextClick(textPane: JTextPane, text: String)
    }

    private class ClickableStyle(
            private val textPane: JTextPane,
            private val text: String,
            private val startOffset: Int,
            private val listener: OnTextClickListener?
    ) {

        private var hover: Boolean = false

        internal fun performClick() {
            listener?.onTextClick(textPane, text)
        }

        internal fun onHover() {
            if (!hover) {
                var attr = updateFontSize(ATTR_EXPLAINS_HOVER, textPane.font, EXPLAINS_FONT_SCALE)
                attr = setClickableStyle(attr, this)
                textPane.styledDocument.setCharacterAttributes(startOffset, text.length, attr, true)

                hover = true
            }
        }

        internal fun clearHover() {
            if (hover) {
                var attr = updateFontSize(ATTR_EXPLAINS, textPane.font, EXPLAINS_FONT_SCALE)
                attr = setClickableStyle(attr, this)
                textPane.styledDocument.setCharacterAttributes(startOffset, text.length, attr, true)

                hover = false
            }
        }

    }

    private class ClickableStyleListener : MouseAdapter() {
        private var mLastHover: ClickableStyle? = null

        override fun mouseClicked(e: MouseEvent) {
            if (e.modifiers and InputEvent.BUTTON1_MASK == 0 || e.clickCount > 1)
                return

            getClickableStyle(e)?.performClick()
        }

        private fun getClickableStyle(e: MouseEvent): ClickableStyle? {
            val textPane = e.component as JTextPane
            val elem = textPane.styledDocument.getCharacterElement(textPane.viewToModel(e.point))
            val attr = elem.attributes as MutableAttributeSet
            return attr.getAttribute(ClickableStyle::class.java) as? ClickableStyle
        }

        override fun mouseMoved(e: MouseEvent) {
            val lastHover = mLastHover
            val hover = getClickableStyle(e)

            if (lastHover != hover) {
                mLastHover = hover

                lastHover?.clearHover()
                hover?.onHover()
            }
        }

        override fun mouseExited(e: MouseEvent) {
            mLastHover?.apply {
                clearHover()
                mLastHover = null
            }
        }
    }

}
